feat: add Python bindings

winstxnhdw opened the pull request:
Now at https://github.com/winstxnhdw/gxhash

winstxnhdw committed:
feat: add gxhash Python bindings

winstxnhdw commented:
Should we consider creating a function with their GIL-less counterpart like `gxhash32_nogil`? Although dropping the GIL is effectively just flipping a boolean variable, our single-threaded users may think that it's an unnecessary performance penalty on them.

But more importantly, is `gxhash` even thread-safe?

winstxnhdw commented:
Perhaps we should also consider moving the `py-gxhash` directory into `ffi/python` and move the existing C FFI to `ffi/C`.

ogxd commented:
Hello @winstxnhdw,

I tried installing py-gxhash via pip using your command but I get this error:
```
ERROR: Could not find a version that satisfies the requirement version-subpkg (unavailable) (from versions: none)
ERROR: No matching distribution found for version-subpkg (unavailable)
```
Am I missing something? I am using python 3.13

> But more importantly, is gxhash even thread-safe?

Gxhash has no shared or global state. Any state is internal to the method scope (thus thread local). The input data is passed by reference, so it must not be mutated while gxhash is reading from it, or the produced hash will be undefined, but it won't crash nor hang. You can indeed drop the GIL if you wish.

winstxnhdw commented:
> Am I missing something? I am using python 3.13

Looks like your shell is not escaping the URL correctly. I have updated the installation instructions. This one should work.

winstxnhdw committed:
feat: add `async` and no GIL functions

winstxnhdw commented:
I notice that the docs uses the specific phrase `arbitrary stream of bytes`, but this is not technically correct, isit? None of the `gxhash*` functions can accept a stream.

winstxnhdw commented:
The current async function is more expensive than running it synchronously even for 5 GB files. Using nogil + threads is more than 20x faster.

Perhaps instead of spawning a tokio thread, we should pass a Python thread to Rust instead.

winstxnhdw commented:
> One remaining issue is that using the hybrid flag does not build on ARM (I am on a Macbook M1) because this feature is reserved for x86 + nightly rust. I changed the behavior to ignore the flag in such cases instead of not building. I also used the no_std crate since Hasher and other stuff is unnecessary here.

Ah, right. I'll be sure to handle that.

> I don't think it makes a lot of sense to expose an async version of gxhash. Async / await is usually reserved for network-bound or sometimes even disk-bound operation so that a thread is not frozen while waiting for the response. Here gxhash is purely CPU-bound. Suppose in a given context the hashing time is substantial enough that you think one may want to execute it asynchronously. In that case, it also means there is a substantial amount of CPU work delegated to a threadpool thread (from tokio or python), which is not something you want (it may cause threadpool starvation). In such case what you want is delegate the work to a dedicated background thread, and you wouldn't use async / await (instead invoke thread and send / consume work via queue for instance).

The reason for exposing an async version is because I'd imagine `gxhash` being used a lot in an asynchronous context. More specifically, in web servers. Running a synchronous function, regardless of whether it has dropped the GIL, within an async route handler **will** block the event loop from handling concurrent requests. If it takes 0.5s for a request to hash a file, that's 0.5s that the server will not be able to respond to any other requests. From how I see it, in this case, there's no difference between how I would handle a synchronous IO-bound task. I would still run it within a separate thread and it would still be susceptible to thread pool starvation.

I am not exactly a software design expert on this topic, so if you still think that it doesn't make sense, let's discuss this further. There's very little resources on this topic, especially in the context of Python and I am always looking to broaden my understanding.

>It is correct, however this documentation is on the Hasher [trait implementation](https://github.com/ogxd/gxhash/blob/eab2961232df5a0908fbe25e44535c11187a2d62/src/hasher.rs#L6) which is a special Rust construct to build a hash pieces by pieces. The mention of arbitrary stream of bytes is actually a copy paste from the official documentation of that trait: https://doc.rust-lang.org/std/hash/trait.Hasher.html

You also used the same description [here](https://github.com/ogxd/gxhash/blob/main/src/gxhash/mod.rs#L22). Anyways, do you ever plan to implement larger-than-memory hashing? After adding `gxhash-py`, I am thinking of implementing two more variants. A streaming variant for larger-than-memory inputs and a parallel variant that'll use `rayon` to hash chunks in parallel. Downside to this, is that the parallel and streaming variants will not emit the same hash as the other variants, given the same input. Not sure if you think this would be poor DX.

> If you like this second option please let me know. That would imply moving transferring this repo to a "gxhash" organization and name it gxhash-rs, transfer gxhash-csharp and then finally gxhash-py.

I don't mind this, but personally, I think there's some beauty and credibility to have a monolith of a library with all the bindings in one place, similar to `polars`. It's a shame that you already made `gxhash-csharp` since that was what I was going to propose next using [csbindgen](https://github.com/Cysharp/csbindgen) which would generate a Rust binding for .NET and Unity without runtime marshalling. I am impartial to how you want to do this as long as it is clearly within the `gxhash` ecosystem, so either in a organisation or within this repository.

ogxd commented:
> The reason for exposing an async version is because I'd imagine gxhash being used a lot in an asynchronous context. More specifically, in web servers. Running a synchronous function, regardless of whether it has dropped the GIL, within an async route handler will block the event loop from handling concurrent requests. If it takes 0.5s for a request to hash a file, that's 0.5s that the server will not be able to respond to any other requests. I have heard other developers mention that even 100 microseconds of blocking is unacceptable. From how I see it, there's no difference between how I would handle a synchronous IO-bound task and a CPU-bound task. As long as neither saturates CPU or memory usage, I would still run it within a separate thread, of which would still be susceptible to thread pool starvation.
> 
> I am not exactly a software design expert on this topic, so if you still think that it doesn't make sense, let's discuss this further. There's very little resources on this topic, especially in the context of Python and I am always looking to broaden my understanding.

Let's take your example: 
We have a backend service handling a request, and at some point, and for some reason while processing that request it must hash some huge data held in memory (let's assume it's not on the disk for now). This operation is purely CPU-bound.  
Now imagine 3 cases:
- The hash is computed synchronously. This means it will effectively block the event loop thread for that duration. We agree it's not ideal.
- The hash is computed asynchronously via async / await. In python, by default, this means the hash computation coroutine will be queue and schedule at some point by the event loop, but still on the event loop thread. So it brings no improvement. 
- The hash is computed asynchronously via async / await, but this time you use a [ThreadPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor) to delegate this work to a pool of threads (called workers in this jargon). Here you're no longer blocking the event loop! However you might have another issue: your operation will block a worker dedicated to the processing of async / await operation for the same duration. This is called _thread pool starvation_ (not to be confused with thread starvation!).
- The hash is computed asynchronously, but not using async / await. Instead, the work is simply delegated to a background thread dedicated to process such heavy operations. This is the way to go.

As a staff backend engineer I do have some experience on this kind of topic, but don't just take my words for it, there are some resources online on this subject. Here are some:
- https://learn.microsoft.com/en-us/dotnet/core/diagnostics/debug-threadpool-starvation#resolving-threadpool-starvation

> To eliminate ThreadPool starvation, ThreadPool threads need to remain unblocked so that they're available to handle incoming work items.

- https://stackoverflow.com/a/71738644/2664927

> Running such operations inside an async def endpoint will block the event loop; and hence, any further client requests will get blocked until the blocking operation is completed.

Now if you read from the disk, or compute a hash block by block, then it's another story. I'd likely use async / await to read from the disk of wait for receiving request bytes, and thus hash on-the-go. In that situation I'd say every chunk hash operation isn't going to be blocking as much as our first example, and thus it might be acceptable to do it synchronously.

> Also, do you have plans to implement larger-than-memory hashing?

In rust, that's one of the purposes of `Hasher` trait. You build a Hasher, you hash pieces by pieces, and then you "finalize" your hash. I'm not sure how easily this structure could be mapped to some python class however using this bindgen method.

winstxnhdw commented:
> The hash is computed asynchronously via async / await, but this time you use a [ThreadPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor) to delegate this work to a pool of threads (called workers in this jargon). Here you're no longer blocking the event loop! However you might have another issue: your operation will block a worker dedicated to the processing of async / await operations for the same duration. This is called thread pool starvation (not to be confused with thread starvation!).

Thank you for taking the time to write something up. But what I am confused about is whether only CPU-bound functions can cause thread pool starvation. In Python, not all IO-bound functions have an async variant. For example, `open` is used to load a file from disk synchronously.

```python
with open("model.bin", "rb") as f:
    file_bytes = f.read()
```

Performing such an operation would block the event loop and one way to avoid blocking the event loop would be to run `open` in another thread. Would this not suffer from thread pool starvation as well? In this case, would you recommend pushing this to a task queue?

In any case, it would be best to remove the `async` variants anyway because it requires a blocking and really expensive copy step to convert `&[u8]` to `Vec<u8>`.

> Oh indeed! We must change that 👍

I will wait for you to change the description and then align my docstrings with yours.

> In rust, that's one of the purposes of Hasher trait. You build a Hasher, you hash pieces by pieces, and then you "finalize" your hash. I'm not sure how easily this structure could be mapped to some python class however using this bindgen method.

I think Python Generators would work well here.

winstxnhdw committed:
feat: remove async variants

winstxnhdw committed:
fix: remove unused imports

winstxnhdw committed:
build: add release profile

mqudsi commented:
> Now if you read from the disk, or compute a hash block by block, then it's another story. I'd likely use async / await to read from the disk or wait for receiving request bytes, and thus hash on-the-go. In that situation I'd say every chunk hash operation isn't going to be blocking as much as our first example, and thus it might be acceptable to do it synchronously.

I would humbly disagree, or at least say that it would only be acceptable with some caveats; this approach would improve p9x latency by blocking the event loop for a shorter amount of time per call, but the total time the event loop remains "unavailable" to handle events would remain the same (well, actually it would increase because of the function call and ffi overhead combined with the decreased gxhash-side performance of hashing smaller chunks at a time) thereby reducing total throughput.

In my opinion though, this isn't something that the library itself should necessarily manage (though I am not a pythonista and I am talking from the general perspective such as that taken in async rust or async C#, both of which I am considerably more familiar with), only because of the number of factors and permutations that have to be taken into account, all of which the downstream caller integrating gxhash into their project would know more about.

The subtleties here lie greatly in the specific application. If you are hashing content that fits comfortably in memory without causing gc issues or if the final content is going to be stored in memory regardless (i.e. is not being streamed to disk or network) then it might make more sense (given how fast gxhash itself is) to buffer the entirety of the content then make one threadpool-backed call to gxhash that doesn't block the event loop, that runs at max gxhash performance, and minimizes the function call/ffi overhead. But if you *are* working on streaming content that won't be stored (contiguously!) in memory thereafter, then you're obviously going to have to call gxhash per some smaller-sized slice at a time. In that case, the question of whether this should be done on a threadpool thread or directly from the async routine is a much more complicated question that would involve knowledge of the size of the input (i.e. if you are hashing 16 bytes at a time, the threadpool overhead, context switch, cache misses, etc is going to exceed the benefits of not blocking the event pool) while if you're processing still decently-sized chunks at a time it might make sense to spin up a threadpool task to handle the cpu-intensive work.

But I don't know what the accepted level of "handholding" here is in the python modules community (or even if python devs are interested in optimizations at this level to begin with, though what I wrote above is assuming at least *some* do).

winstxnhdw commented:
> But I don't know what the accepted level of "handholding" here is in the python modules community (or even if python devs are interested in optimizations at this level to begin with, though what I wrote above is assuming at least some do).

I always believed that good API design should include sane defaults while also giving the user the option to build their own specialised solution from scratch, and FWIW, I care deeply about such optimisations. I also think providing the user with an `async` variant is better DX but I can also see that it may mislead users to think that the `async` variant is most optimal for their use case.

> In that case, the question of whether this should be done on a threadpool thread or directly from the async routine is a much more complicated question that would involve knowledge of the size of the input (i.e. if you are hashing 16 bytes at a time, the threadpool overhead, context switch, cache misses, etc is going to exceed the benefits of not blocking the event pool)

I am not sure what could be much worse than blocking the event loop. When the event loop is blocked, no one can load your web page, your Prometheus service will no longer be able to scrape for metrics, and your monitoring service sends an alert to everyone that your service is down. Of course the other option is for users to push it to a task queue, but on a sufficiently large cluster with a known amount of users, is there really a need to force the user to pay the development complexity and overhead cost of using a task queue?

mqudsi commented:
> I am not sure what could be much worse than blocking the event loop.

The implication is that you are still blocking the event loop while it does the tasks I mentioned. e.g. one heavily optimized ffi call to gxhash to hash a 4-byte input might actually block the event loop less than the bookkeeping it has to do (synchronously) to send and retrieve the result to an off-thread worker for such a low-latency operation. And if you expand the scope past microbenchmarks, you need to take into account whether or not the "other thread" will cohabit a core that's already servicing another even loop for your application, the trashing of multiple threads' L1 cache lines, etc.

winstxnhdw commented:
In the context of Python, once you cross the FFI boundary and drop the GIL, everything from then on is effectively non-blocking. Unless you mean the hashing operation saturates the CPU.

winstxnhdw committed:
feat/perf: add performant async implementation

winstxnhdw committed:
feat: use class API instead

winstxnhdw committed:
docs: improve docstring clarity

winstxnhdw committed:
docs: keep docstrings consistent

winstxnhdw committed:
fix: propagate mmap errors to Python

winstxnhdw commented:
This PR is finally ready but it seems like the `gxhash` package on PyPI is being taken by @michalgolda. Michal, if it is possible, can we ownership of the name?

winstxnhdw committed:
docs: describe perfomance quirk at large input sizes

winstxnhdw committed:
docs: add docstrings for each hasher class

winstxnhdw commented:
Hey @ogxd, thinking back about our earlier conversation, and due to the existence of `gxhash-csharp`, perhaps we should create an organisation for `gxhash` and I'll maintain the Python bindings.

winstxnhdw committed:
docs: populate README

winstxnhdw committed:
chore: remove redundant `drop`

winstxnhdw committed:
feat: replace `hash_nogil` with `hash_async`

winstxnhdw committed:
docs/build: bump version

winstxnhdw committed:
fix: do not install `gxhash` as relative path

winstxnhdw committed:
build/docs: disable `hybrid` feature by default

winstxnhdw committed:
fix: use `PyBytes` to reference bytes in `future_into_py`
- credits to kevin reid

winstxnhdw committed:
feat!: accept file path instead of file

winstxnhdw committed:
docs: update file hashing examples

winstxnhdw committed:
docs: remove `nogil` feature

winstxnhdw committed:
docs: recommend installing with `hybrid`

kevinjohncutler commented:
@ogxd @winstxnhdw what is the current status here? I've been using the pypi package with no issues on macOS and Linux, but on attempting to pip install gxhash on Windows, the build is failing during the rust compilation:

```
Compiling py-gxhash v0.1.7 (C:\Users\runneradmin\AppData\Local\Temp\pip-install-1wo9bpwb\gxhash_1d4765deac6943f5917b44f18cd3692c)
  error[E0432]: unresolved import `std::os::fd`
   --> src\lib.rs:3:14
    |
  3 | use std::os::fd::FromRawFd;
    |              ^^ could not find `fd` in `os`
    |
  note: found an item that was configured out
   --> /rustc/f8297e351a40c1439a467bbbb6879088047f50b3\library\std\src\os\mod.rs:185:10
    |
    = note: the item is gated here
  
  error[E0599]: no function or associated item named `from_raw_fd` found for struct `File` in the current scope
    --> src\lib.rs:14:60
     |
  14 |     let mmap = unsafe { memmap2::Mmap::map(&std::fs::File::from_raw_fd(libc::dup(file_descriptor)))? };
     |                                                            ^^^^^^^^^^^ function or associated item not found in `File`
     |
  note: if you're trying to build a new `File` consider using one of the following associated functions:
        File::open
        File::create
        File::create_new
    --> /rustc/f8297e351a40c1439a467bbbb6879088047f50b3\library\std\src\fs.rs:452:5
  help: there is an associated function `from_raw_handle` with a similar name
     |
  14 -     let mmap = unsafe { memmap2::Mmap::map(&std::fs::File::from_raw_fd(libc::dup(file_descriptor)))? };
  14 +     let mmap = unsafe { memmap2::Mmap::map(&std::fs::File::from_raw_handle(libc::dup(file_descriptor)))? };
     |
  
  Some errors have detailed explanations: E0432, E0599.
  For more information about an error, try `rustc --explain E0432`.
  error: could not compile `py-gxhash` (lib) due to 2 previous errors
  💥 maturin failed
    Caused by: Failed to build a native library through cargo
    Caused by: Cargo build finished with "exit code: 101": `"cargo" "rustc" "--profile" "release" "--features" "pyo3/extension-module" "--message-format" "json-render-diagnostics" "--manifest-path" "C:\\Users\\runneradmin\\AppData\\Local\\Temp\\pip-install-1wo9bpwb\\gxhash_1d4765deac6943f5917b44f18cd3692c\\Cargo.toml" "--lib"`
  Error: command ['maturin', 'pep517', 'build-wheel', '-i', 'C:\\hostedtoolcache\\windows\\Python\\3.10.11\\x64\\python.exe', '--compatibility', 'off'] returned non-zero exit status 1
  [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for gxhash

```
It appears the crate relies on Unix-specific file descriptor traits (std::os::fd) which are not available in the Windows standard library (which uses std::os::windows::io).

winstxnhdw commented:
Thanks for the feedback. I'll remove the `hash_file` API then. I don't think anyone uses it. I added it initially to avoid the implicit slow copy of `Py<Bytes>` to `u8`, but I have since found a better way to do this. I'll push the new version in a bit.

winstxnhdw committed:
feat: release v0.2.2
- mark class as frozen and immutable
- remove `hash_file*` APIs
- fix windows compat
- update PyO3 API
- turn off all default features
- use fat lto
- use 2024 rust edition

winstxnhdw commented:
I've release v0.2.2 which should work on Windows now. Can't run CI until @ogxd merges.

winstxnhdw committed:
perf: inline hasher

winstxnhdw committed:
feat!: make `bytes` parameter positional-only

winstxnhdw committed:
feat: bump minor version

winstxnhdw committed:
chore: remove redundant `Ok`

ogxd commented:
Hello @winstxnhdw 

I can see that you've put a lot of efforts into wrapping gxhash in python, and it seems like other people are interested as well. That said, I think we should keep this crate about gxhash in rust only for simplicity. I can't think of a scenario where I'd bring any value if we create a gxhash organization, but it's perfectly fine to me if you create your own `py-gxhash` repository (or `gxhash-py`, call it however you want). I'll do my best to makes changes to the gxhash rust code if, in the future, there are changes of general interest that could profit `py-gxhash` as well.

(I've allowed the github workflows to run in the meantime)

winstxnhdw commented:
Alright, no worries. I'll close this PR once I've moved this to my own repo. Thanks for the help! Do let me know if you ever change your mind.
